Feature/4.8.1 client steering mandate

rmelotte opened the pull request:
The source BSSID was missing from the BTM report, which made test 4.8.1 fail.

Add it to the internal BTM responses so that it can also be added to the BTM report.

Fixes #429

rmelotte committed:
common: sNodeBssSteerResponse: add source_bssid and target_bssid

When a BTM response is received, the MultiAP standard specifies that
the agent should send a BTM report back to the controller. That BTM
report has to contain the source BSSID.

Since the standard BTM response sent by the client doesn't contain the
source bssid, there is no way for the slave_thread to add it to the
BTM report.

When receiving the BTM response we know on which VAP it's received, so
we can just add the source bssid to the BTM response used internally.

Add the source_bssid in sNodeBssSteerResponse so that each BWL can add
it when receiving a BTM reponse.

While we're at it, we know that the BTM response also contains an
optional target_bssid that should also be used in the BTM report, so
add it as well.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
agent: ap_manager_thread: include the source bssid in BTM response

Now that the BWL includes the source bssid in the BTM response, we can
include it as well in the BTM response that is going to be sent to the
slave thread.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
bwl: dummy: include the BSSID in the BTM response

Now that the BTM response has a field for the BSSID, we can include it
so that son_slave_thread can later use it.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
bwl: dwpal: include the BSSID in the BTM response

Now that the BTM response has a field for the BSSID, we can include it
so that son_slave_thread can later use it.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
agent: slave: add source BSSID in BTM report

Now that the ap manager includes the source BSSID in the BTM response,
we can include it when sending a BTM report.

This is the last step need needed for the flow 4.8.1.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte commented:
The test `MAP-4.8.1_ETH_FH24G` on rax40 is here: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/449752655 (using the Marvell controller, and not Broadcom).

rmelotte commented:
The test 4.8.1 on rax40 is here: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/449752655

rmelotte commented:
I took a shortcut when I said it allows us to pass 4.8.1 .
With this PR we're passing test `MAP-4.8.1_ETH_FH24G` but with a different controller than what the test plan says (until https://github.com/prplfoundation/prplMesh/issues/886 is fixed).

rmelotte commented:
> The reserved fields are there for alignment, which is messed up now with the bssids. I think the status code should come at the end so the others are more or less aligned. And reserved3 I would just remove, it's pretty pointless at the end of a struct.

I don't really understand... What were the 3 reserved uint8_t supposed to do before this change?
I changed it to have status_code at the end.

morantr commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -2276,22 +2276,25 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
-        if (dwpal_string_to_struct_parse(buffer, bufLen, fieldsToParse, sizeof(MACAddress)) ==
-            DWPAL_FAILURE) {
+        if (dwpal_string_to_struct_parse(buffer, bufLen, fieldsToParse,
+                                         sizeof(fieldsToParse)) == DWPAL_FAILURE) {
```
Not sure why, but I have a Deja Vu when I wrote this comment.
Anyway, setting the last argument to `sizeof(fieldsToParse)` is wrong.
According to DWPAL code, the last argument should is defined by this description:
`size_t userBufLen - The user total buffer size to hold parsed info from inside fieldsToParse`
This means that it actually should be the size of `vap_name + MACAddress + status_code`.

I see wrong usage of it on all DWPAL functions.

arnout replied:
I wondered about that as well...

rmelotte replied:
Replaced with the sum of the size of each element.

morantr reviewed (approved):
Only one comment, but approved since it is a small issue.

arnout commented:
> > The reserved fields are there for alignment, which is messed up now with the bssids. I think the status code should come at the end so the others are more or less aligned. And reserved3 I would just remove, it's pretty pointless at the end of a struct.
> 
> I don't really understand... What were the 3 reserved uint8_t supposed to do before this change?
> I changed it to have status_code at the end.

They were supposed to make sure the structure is aligned. However, it was already pretty bogus. sMacAddr is properly declared as packed, so it will occupy exactly 6 bytes. However, since sNodeBssSteerResponse is not declared as packed, the compiler may already insert 2 additional bytes between `mac` and `status_code`. So the only effect of the three reserved fields is to make sure the whole thing is 2-byte aligned. Which is anyway pointless, because it doesn't have any fields that require alignment. And moreover, in other places (e.g. tlvf) we're using structures in a way that assume the CPU has no alignment restrictions at all...

I think those reserved fields are simply cargo cult.

So my proposal would be to either insert a preparatory commit that removes all the reserved fields from base_wlan_hal_types.h, or to remove them only in sNodeBssSteerResponse in this commit.

arnout reviewed (approved):
I gave my review comments on the phone - sorry about that, it really sucks...

Some review comments really have to be taken into account, but I'm approving already with don't merge label.

rmelotte committed:
controller: rename btm_report_received events

When receiving a BTM report, the logs would print 'BSS_TM_RESP from
client_mac', which is not only confusing, it also prevents us from
checking for the btm report in the log files.

Change the log message to use BTM_REPORT instead of BSS_TM_RESP.
Also include the source bssid in the logs.

Renamed the BSS_TM_RESPONSE_RECEIVED to BTM_REPORT_RECEIVED to avoid
confusion.

Similarly, rename bool bss_tm_response_received to
btm_report_received.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tests: client_steering_mandate: check the BTM report on the controller

Since in the end the BTM report is for the controller, check that it
received it (with the correct source bssid).

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte commented:
Unfortunately some of the comments disappeared with the force-push, but the only part I didn't address yet is this one:
`beerocks::net::network_utils::mac_from_string(get_radio_mac());`

The relevant comment is here: https://github.com/prplfoundation/prplMesh/pull/891#commitcomment-37577580
